### PR TITLE
Fix nczarr-experimental: improve build support, disengage hdf5 vs netcdf4 flags, and find AWS libraries

### PR DESCRIPTION
**Pull request by DennisHeimbigner**
The primary fix is to improve CMake build support.
Specific changes include:
* CMake: Provide a better soln to locating the AWS SDK libraries; the new way is the preferred method as described in the aws-cpp-sdk documentation.
* CMake (and Automake): allow -DENABLE_S3_SDK (default off) to suppress  looking for AWS libraries.
* CMake: add the complete set of nczarr tests
* CMake: add EXTERNL as needed to various .h files.
* Improve support for windows drive letters in paths.
* Add nczarr and s3 flags to nc-config
* For VisualStudio X nczarr, cleanup the NAN+INFINITY handling
* Convert _MSC_VER -> _WIN32 and vice versa as needed
* NCZarr - support multiple platform paths including windows, cygwin, mingw, etc.
* NCZarr - sort the test outputs because different platforms produce directory contents in different orders.

One big change concerns netcdf-c/CMakeLists.txt and netcdf-c/configure.ac. In the current versions, it was the case that --disable-hdf5 disabled netcdf-4 (libsrc4). With nczarr, this can no longer be the case because nczarr requires libsrc4 even if libhdf5 is disabled. So, I modified the above files to move the format options (HDF5, NCZarr, HDF4, etc) to a single place near the front of the files. Now it is the case that:
* Enabling any of the formats that require libsrc4 also does an implicit --enable-netcdf4.
* --disable-netcdf4 | --disable-netcdf-4 now becomes an alias for --disable-hdf5.

There are probably some bugs in this change in terms of dependencies between format options.

Problems:
* CMake S3 support is still not working for Visual Studio. The code will build, but the S3 tests will not execute.
* A recent issue points out that there is work to do on handling UTF8 filenames, but that will be addressed in a separate fix.

Notes:
* Consider converting all of our includes/.h files to use EXTERNL

### CODE REVIEW

File: include/nc4internal.h [hunk 1]
```diff
@@ -338,100 +338,100 @@ typedef struct
 } nc_hvl_t;
 
 /* Misc functions */
-int NC4_inq_atomic_type(nc_type typeid1, char *name, size_t *size);
-int NC4_lookup_atomic_type(const char *name, nc_type* idp, size_t *sizep);
+extern int NC4_inq_atomic_type(nc_type typeid1, char *name, size_t *size);
```
Comment (captainkirk99) [inline include/nc4internal.h:341]: Aren't all functions exern by default?

**Comment by WardF** [comment]
Fails to build on OSX.  Need to fencepost malloc.h in zincludes.h:23 with #ifdef HAVE_MALLOC_H.  @DennisHeimbigner I can add this on a separate branch if you prefer, let me know.

**Comment by WardF** [comment]
Compilation failure on OSX is gone, but seeing a test failure now.  More details to follow.

**Comment by WardF** [comment]
Observing the following:

````
wfisher@kashian:~/Desktop/tmp/netcdf-c/build/nczarr_test$ ./run_it_test1.sh 
*** Testing ncgen with -4 and zmap=nz4

*** Test cycle zext=nz4
*** SUCCEED: dimscope
*** SUCCEED: tst_group_data
*** SUCCEED: tst_solar_1
7c7
< 		fvar:att = -Infinityf, NaNf, Infinityf ;
---
> 		fvar:att = Infinityf, NaNf, Infinityf ;
10c10
< 		dvar:att = -Infinity, NaN, Infinity ;
---
> 		dvar:att = Infinity, NaN, Infinity ;
````

**Comment by DennisHeimbigner** [comment]
With respect to the infinity failure, it appears that the expression
> -INFINITY
is being treated like
> +INFINITY
If you look at the file include/isnan.h it assumes that INFINITY is defined.
You might look at the build output and see if there are any warning or such
that mention INFINITY or NAN.

**Comment by DennisHeimbigner** [comment]
Added commits to elide PARALLEL code in libnczarr.
Also added definitions of isnan and isinf in an attempt
to solve the OSX infinity problem.

**Pull request merged**